### ISSUE TITLE
Horizontal and vertical Flip of video?

### ISSUE THREAD

**Issue by tonfa80** [moritzmhmk/homebridge-camera-rpi#4]
Hi,

I am new with ffmpeg tool and I am wondering how can I flip Raspberry pi camera image horizontally and vertically in home app?
I've found and edited "_/usr/local/lib/node_modules/homebridge-camera-rpi/CameraSource.js_" file and added "_-vf hflip,vflip_" on ffmpeg command on '_Camera.prototype.handleSnapshotRequest_' function and it works on preview image on home app.
I understood that similarly  "_-vf hflip,vflip_" should be added into  '_Camera.prototype.handleStreamRequest_' function as well. However, ffmpeg complains error that:

> Filtergraph 'hflip,vflip' wa defined for video output stream 0:0 but codec copy was selected. Filtering and streamcopy cannot be used together

I wonder how this issue should be handled?

**Comment by moritzmhmk** [comment]
The good performance of this camera implementation is achieved by letting the GPU (instead of the CPU) do most of the job (i.e. setting the size and format via the video4linux2 driver and only using ffmpeg to copy the frames over into a valid srtp stream).

Using a filter (-vf) is not a good idea. Instead one should make the driver (vide4linux2) do the flipping or rotating. FFmpeg only supports setting some of the v4l2 driver options and rotation and flipping is none of them. So instead you need to use `v4l2-ctl` to set those.

TL;DR
try `v4l2-ctl --set-ctrl vertical_flip=1` or `v4l2-ctl --set-ctrl=rotate=180`

**Comment by tonfa80** [comment]
Thanks. It works now.

Toni

**Comment by charlietomo** [comment]
For the non-programmer, are you able to point out exactly where you put the :

>v4l2-ctl --set-ctrl vertical_flip=1 

I can find line 164 and that region in the code, but my first two edits failed. Many thanks in advance.

**Comment by tonfa80** [comment]
I have set it as follows on line 164:
`let v4l2ctlCommand = `--set-ctrl=rotate=180,video_bitrate=${bitrate}000`

**Comment by moritzmhmk** [comment]
You can execute `v4l2-ctl --set-ctrl vertical_flip=1` in the terminal and it will cause the camera to be flipped - adding this to the code is not necessary. I will add this as a config option in the next release.

**Comment by charlietomo** [comment]
> You can execute v4l2-ctl --set-ctrl vertical_flip=1 in the terminal and it will cause the camera to be flipped - adding this to the code is not necessary. I will add this as a config option in the next release.
Thanks that is much better than editing code manually, look forward to the config option.

**Comment by pokono** [comment]
+1

**Comment by moritzmhmk** [comment]
added options (0f6802c) to config
